lib/active_job/queue_adapters/solid_queue_ext.rb

```diff
@@ -40,7 +40,7 @@ def supported_job_statuses
   end
 
   def supported_job_filters(*)
-    [ :queue_name, :job_class_name ]
+    [ :queue_name, :job_class_name, :finished_at_start, :finished_at_end ]
```

Comment:
I think we want `finished_at` here 🤔

Reply:
Ah, no, this had to be split into `supported_natively` and `supported_in_memory`, but in any case, it should be just `finished_at` when added to `supported_natively`.

Reply:
I had something like this in mind:
```diff
diff --git a/lib/active_job/jobs_relation.rb b/lib/active_job/jobs_relation.rb
index e4e19ee..a92af46 100644
--- a/lib/active_job/jobs_relation.rb
+++ b/lib/active_job/jobs_relation.rb
@@ -23,7 +23,6 @@ class ActiveJob::JobsRelation
   include Enumerable
                                                                                                                                      
   STATUSES = %i[ pending failed in_progress blocked scheduled finished ]
-  FILTERS = %i[ queue_name job_class_name finished_at ]
                                                                                                                                      
   PROPERTIES = %i[ queue_name status offset_value limit_value job_class_name worker_id recurring_task_id finished_at ]
   attr_reader *PROPERTIES, :default_page_size
@@ -276,8 +275,7 @@ class ActiveJob::JobsRelation
     end
                                                                                                                                      
     def filters
-      # @filetrs ||= queue_adapter.supported_job_filters(self).select { |property| public_send(property).present? && !queue_adapter.natively_supported_job_filters?(self, property) }
-      @filters ||= FILTERS.select { |property| public_send(property).present? && !queue_adapter.supports_job_filter?(self, property) }
+      @filters ||= queue_adapter.supported_job_filters(self).select { |property| public_send(property).present? && !queue_adapter.natively_supported_job_filters(self).include?(property) }
     end
                                                                                                                                      
     def ensure_failed_status
diff --git a/lib/active_job/queue_adapters/resque_ext.rb b/lib/active_job/queue_adapters/resque_ext.rb
index 17b2fbf..d7d28af 100644
--- a/lib/active_job/queue_adapters/resque_ext.rb
+++ b/lib/active_job/queue_adapters/resque_ext.rb
@@ -48,6 +48,10 @@ module ActiveJob::QueueAdapters::ResqueExt
   end
                                                                                                                                      
   def supported_job_filters(jobs_relation)
+    [ :queue_name, :job_class_name ]
+  end
+
+  def natively_supported_job_filters(jobs_relation)
     if jobs_relation.pending? then [ :queue_name ]
     else []
     end
diff --git a/lib/active_job/queue_adapters/solid_queue_ext.rb b/lib/active_job/queue_adapters/solid_queue_ext.rb
index 2db9ea1..b069504 100644
--- a/lib/active_job/queue_adapters/solid_queue_ext.rb
+++ b/lib/active_job/queue_adapters/solid_queue_ext.rb
@@ -39,8 +39,14 @@ module ActiveJob::QueueAdapters::SolidQueueExt
     SolidQueueJobs::STATUS_MAP.keys
   end
                                                                                                                                      
-  def supported_job_filters(*)
-    [ :queue_name, :job_class_name, :finished_at_start, :finished_at_end ]
+  def supported_job_filters(jobs_relation)
+    filters = [ :queue_name, :job_class_name ]
+    filters << :finished_at if jobs_relation.finished?
+    filters
+  end
+
+  def natively_supported_job_filters(*)
+    [ :queue_name, :job_class_name, :finished_at ]
   end
                                                                                                                                      
   def jobs_count(jobs_relation)

```

Reply:
I was thinking about raising query errors if you try to use unsupported filters

Reply:
I think `natively_supported_job_filters` should take a `job_relation` as well and do the check `jobs_relation.finished?`. Then you just use that in `supported_job_filters` 🤔

Reply:
a little recap to see if we both are in the same page:
given that `JobsRelations` is the interface to the adapters and therefore knows all the generic filters that can be applied: `FILTERS` (with the extra parametrs in where method), I think the most reasonable way to `select` the filters that should be performed in memory is to ask the adapter if the filter is supported natively given the `JobsRelation` state it's not supported natively.

translating this thoughts to code I was thinking something like this:
```ruby
def filters
  @filters ||= FILTERS.select do |property|
    public_send(property).present? && should_perform_filter_in_memory?(filter)
  end
end

def should_perform_filter_in_memory?(filter)
  # example: finished_at is supported only if the status is finished
  queue_adapter.supports_job_filter?(self, property) &&
    !queue_adapter.natively_supported_job_filters?(self).include?(property)
end
```

Reply:
I think it'd make more sense to replace `FILTERS` with `queue_adapter.supported_filters?(self)`, and filter out the ones that aren't supported natively 🤔 The reason is that now the adapter needs to know about all filters anyway, as it needs to respond to `supports_job_filters?`.

Reply:
Yes, that's true 🤔  Hmm.. let me think about it a bit more. I don't like the asymmetry of all these methods right now. Let me try to come up with something by looking at the whole thing.

Reply:
Ok, ok. This can be all made much simpler. We don't need `natively_supported_..` filters or anything like that. We can just keep the same implementation as before because we aren't checking that to show filters in `_filters.html.erb`. This is what I suggest, after looking at the whole thing: 

We keep this for Resque: 
```ruby
  def supported_job_filters(jobs_relation)
    if jobs_relation.pending? then [ :queue_name ]
    else []
    end
  end
```

We do this for Solid Queue: 
```ruby
  def supported_job_filters(*)
    [ :queue_name, :job_class_name, :finished_at ]
  end
```
And then, in `JobRelation` we do: 

```ruby
    def filters
      @filters ||= applicable_filters.select { |property| public_send(property).present? && !queue_adapter.supports_job_filter?(self, property) }
    end
```

And we keep `JobRelation`without changing. It doesn't need to change. If someone were to use `finished_at` with Resque (or even Solid Queue with a set of jobs that aren't finished), what would happen is that the result would be empty, which is correct! 😅 

Sorry for the back-and-forth, sometimes, it's hard to see the whole picture when looking at a PR, as the existing implementation kind of anchors you 😅 #209 will introduce some other filters by date that will layer very nicely on top of this without changing anything else.

Reply:
Sorry, I said we don't need to change `JobRelation`, but we need to add `finished_at` to `PROPERTIES`.

Reply:
no worries :)

sounds good, how did you intended to implement `applicable_filters` ?

Reply:
Oops, no! Sorry, I forgot to delete that from my comment! It was part of my previous thoughts, before I realised we didn't need it! Going to edit my comment to avoid confusion.